### PR TITLE
bootspec fixups

### PR DESCRIPTION
**Pull request by AdrianVovk**
Various fixes for the bootspec util-lib. See commit messages for more details.

Split out from #34767

### CODE REVIEW

**Comment by yuwata** [comment]
```
../src/shared/bootspec.c: In function ‘boot_entries_find_unified_global_addons’:
../src/shared/bootspec.c:1165:81: error: passing argument 5 of ‘boot_entries_find_unified_addons’ from incompatible pointer type [-Werror=incompatible-pointer-types]
 1165 |         return boot_entries_find_unified_addons(config, dirfd(d), d_name, root, &ret_addons);
      |                                                                                 ^~~~~~~~~~~
      |                                                                                 |
      |                                                                                 BootEntryAddons **
../src/shared/bootspec.c:1087:34: note: expected ‘BootEntryAddons *’ but argument is of type ‘BootEntryAddons **’
 1087 |                 BootEntryAddons *ret_addons) {
      |                 ~~~~~~~~~~~~~~~~~^~~~~~~~~~
../src/shared/bootspec.c: In function ‘boot_entry_to_json’:
../src/shared/bootspec.c:1961:29: error: passing argument 2 of ‘json_cmdline’ from incompatible pointer type [-Werror=incompatible-pointer-types]
 1961 |         r = json_cmdline(e, &c->global_addons, opts, &v);
      |                             ^~~~~~~~~~~~~~~~~
      |                             |
      |                             const BootEntryAddons (*)[2]
../src/shared/bootspec.c:1754:40: note: expected ‘const BootEntryAddons *’ but argument is of type ‘const BootEntryAddons (*)[2]’
 1754 |                 const BootEntryAddons *global_arr,
      |                 ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
cc1: all warnings being treated as errors
```

File: src/shared/bootspec.c [hunk 1]
```diff
@@ -1811,7 +1843,7 @@ int show_boot_entry(
 
                 const char *p = e->root ? path_startswith(e->path, e->root) : NULL;
                 if (p) {
-                        text = strjoin(ansi_grey(), e->root, "/", ansi_normal(), "/", p);
+                        text = strjoin(ansi_grey(), e->root, "/", ansi_normal(), p);
```
Comment (YHNdnzj) [inline src/shared/bootspec.c]: This is intentional and double / is printed consistently for paths under $BOOT. See comment in boot_entry_file_list for details. Hence please drop this change

**Review by YHNdnzj — CHANGES_REQUESTED** [review]

**Comment by yuwata** [comment]
fuzz-bootspec is unhappy.

File: src/shared/bootspec.c [hunk 2]
```diff
@@ -426,7 +448,9 @@ void boot_config_free(BootConfig *config) {
         FOREACH_ARRAY(i, config->entries, config->n_entries)
                 boot_entry_free(i);
         free(config->entries);
-        free(config->global_addons.items);
+
+        FOREACH_ARRAY(i, config->global_addons, _BOOT_ENTRY_SOURCE_MAX)
```
Comment (YHNdnzj) [inline src/shared/bootspec.c:452]: FOREACH_ELEMENT

File: src/shared/bootspec.h [hunk 3]
```diff
@@ -93,6 +101,9 @@ typedef struct BootConfig {
 const char* boot_entry_type_to_string(BootEntryType);
 const char* boot_entry_type_json_to_string(BootEntryType);
 
+const char* boot_entry_source_to_string(BootEntrySource);
+const char* boot_entry_source_json_to_string(BootEntrySource);
```
Comment (YHNdnzj) [inline src/shared/bootspec.h:105]: mark as `_const_`/`_pure_`?

**Pull request merged**